components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/service/IdentityDataStoreServiceImpl.java

```diff
@@ -0,0 +1,158 @@
+            "org.wso2.carbon.identity.governance.store.UserStoreBasedIdentityDataStore";
+
+    public IdentityDataStoreServiceImpl() throws ClassNotFoundException, InstantiationException,
+            IllegalAccessException {
+
+        String storeClassName = IdentityUtil.getProperty(IDENTITY_DATA_STORE_TYPE);
+        if (DEFAULT_JDBC_IDENTITY_DATA_STORE.equals(storeClassName)) {
+            identityDataStore =  new JDBCIdentityDataStore();
+        } else if (DEFAULT_USER_STORE_BASED_IDENTITY_DATA_STORE.equals(storeClassName)) {
+            identityDataStore = new UserStoreBasedIdentityDataStore();
+        } else {
+            Class clazz = Class.forName(storeClassName.trim());
+            identityDataStore = (UserIdentityDataStore) clazz.newInstance();
+        }
+    }
+
+    @Override
+    public boolean storeInIdentityDataStore(String userName, UserStoreManager userStoreManager, String operationType,
+                                            Map<String, String> claims) throws UserStoreException {
+
+        // Check if the user is locked.
+        if (PRE_SET_USER_CLAIM_VALUES.equals(operationType)) {
+            boolean accountLocked = Boolean.parseBoolean(claims.get(UserIdentityDataStore.ACCOUNT_LOCK));
+            if (accountLocked) {
+                IdentityErrorMsgContext customErrorMessageContext = new IdentityErrorMsgContext(UserCoreConstants
+                        .ErrorCode.USER_IS_LOCKED);
+                IdentityUtil.setIdentityErrorMsg(customErrorMessageContext);
+            }
+        }
+
+        // No need to separately handle if data identityDataStore is user store based.
+        if (isUserStoreBasedIdentityDataStore() || isStoreIdentityClaimsInUserStoreEnabled(userStoreManager)) {
+            return true;
+        }
+
+        // Top level try and finally blocks are used to unset thread local variables.
+        try {
+            if (!IdentityUtil.threadLocalProperties.get().containsKey(operationType)) {
+                IdentityUtil.threadLocalProperties.get().put(operationType, true);
+
+                UserIdentityClaim userIdentityClaim = null;
+                if (!StringUtils.equalsIgnoreCase(operationType, PRE_USER_ADD_CLAIM_VALUES)) {
+                    // We avoid loading claims for pre user add operations.
+                    userIdentityClaim = identityDataStore.load(userName, userStoreManager);
+                }
+
+                if (userIdentityClaim == null) {
+                    userIdentityClaim = new UserIdentityClaim(userName);
+                }
+
+                Iterator<Map.Entry<String, String>> claimsIterator = claims.entrySet().iterator();
+
+                while (claimsIterator.hasNext()) {
+                    Map.Entry<String, String> claim = claimsIterator.next();
+                    String key = claim.getKey();
+                    String value = claim.getValue();
+                    if (key.contains(UserCoreConstants.ClaimTypeURIs.IDENTITY_CLAIM_URI_PREFIX)) {
```

Comment:
this is prone to NPEs

Reply:
Fixed with https://github.com/wso2-extensions/identity-governance/pull/717/commits/21ecba47eebc28b5b1490d86f27de56db0eb78cf